numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -2274,14 +2274,18 @@ array_count_nonzero(PyObject *NPY_UNUSED(self), PyObject *const *args, Py_ssize_
         return NULL;
     }
 
-    count =  PyArray_CountNonzero(array);
-
+    count = PyArray_CountNonzero(array);
     Py_DECREF(array);
 
     if (count == -1) {
         return NULL;
     }
-    return PyLong_FromSsize_t(count);
+
+    PyArray_Descr *descr = PyArray_DescrFromType(NPY_INTP);
+    if (descr == NULL) {
+        return NULL;
+    }
+    return PyArray_Scalar(&count, descr, NULL);
```

Comment:
This is a more significant change than it may look, since it has serious impact on promotion for `count_nonzero` without an axis.

Maybe we can do it, but we should discuss it briefly/add a release note for visibility.
But I am tempted to fix it in the array api tests to say that it is completely fine to return an integer for `count_nonzero(arr, axis=None)`.

(I think an integer return is just better for NumPy users, the argument against it is only that we can also return arrays of course, for which there is no equivalent behavior obviously.)

Reply:
Sure! I can just skip this test or we can keep this change, I'm Ok with both. I added it to today's triage meeting for broader discussion. I can't attend myself today so just ping me if anything was decided.

Reply:
Thanks, I just checked and NumPy doesn't really run into this (I suppose we don't really have code paths that never pass an `axis`, so have to provision anyway).
Still think we should at least mention it in a release note as a subtle change, though.

Reply:
Yeah, I understand that consistency is better with the change.  But in contexts where axis is always `None`, the integer return is more useful and changing it can change results, because:
```
arr = np.linspace(0, 100, 10000, dtype=np.float32)
res = arr / np.count_nonzero(arr)
```
will change from being a `float32` result to a `float64` one.

Not that I suspect this to be seen often.  `skimage` has a function that will return a `float64` rather than a Python float with this change for example, I am sure that usually doesn't matter.